### PR TITLE
support numactl on Linux in multi-threading 2.0

### PR DESCRIPTION
**Pull request by wangleis**
### Details:
 - *Check valid CPUs and update CPU map table on Linux*

### Tickets:
 - *CVS-110073*

### CODE REVIEW

**Comment by wangleis** [comment]
This PR will focus on numactl support on Linux. 

CPU map refactor for Linux will move to https://github.com/openvinotoolkit/openvino/pull/17450

File: src/inference/src/os/lin/lin_system_conf.cpp [hunk 1]
```diff
@@ -43,13 +43,85 @@ CPU::CPU() {
...
+                _proc_type_table[0][ALL_PROC]++;
+                _proc_type_table[0][row[CPU_MAP_CORE_TYPE]]++;
+                if (_proc_type_table.size() > 1) {
+                    _proc_type_table[row[CPU_MAP_SOCKET_ID] + 1][ALL_PROC]++;
+                    _proc_type_table[row[CPU_MAP_SOCKET_ID] + 1][row[CPU_MAP_CORE_TYPE]]++;
+                }
+            }
+
+            if (_proc_type_table.size() > 1) {
+                long unsigned int n = 0;
```
Comment (riverlijunjie) [inline src/inference/src/os/lin/lin_system_conf.cpp:94]:
```suggestion
                size_t n = 0;
```

File: src/inference/src/os/lin/lin_system_conf.cpp [hunk 2]
```diff
@@ -43,13 +43,85 @@ CPU::CPU() {
...
+
+                while (n < _proc_type_table.size()) {
+                    if (0 == _proc_type_table[n][ALL_PROC]) {
+                        _proc_type_table.erase(_proc_type_table.begin() + n);
+                        n--;
```
Comment (riverlijunjie) [inline src/inference/src/os/lin/lin_system_conf.cpp:99]: Could it guarantee `n>0` and then `n--`? Maybe some strange platform may hit it, so we need protect it from this case.

File: src/inference/src/os/lin/lin_system_conf.cpp [hunk 3]
```diff
@@ -43,13 +43,85 @@ CPU::CPU() {
...
+                total_proc++;
+                cores_list.emplace_back(_cpu_mapping_table[i][CPU_MAP_CORE_ID]);
+                valid_cpu_mapping_table.emplace_back(_cpu_mapping_table[i]);
+                if (_cpu_mapping_table[i][CPU_MAP_CORE_TYPE] == MAIN_CORE_PROC) {
+                    phy_core_list.emplace_back(_cpu_mapping_table[i][CPU_MAP_GROUP_ID]);
+                }
+            }
+        }
+
+        if (total_proc == _processors) {
```
Comment (riverlijunjie) [inline src/inference/src/os/lin/lin_system_conf.cpp:70]: Is it considered the possible of `total_proc==0` ?

File: src/inference/src/os/lin/lin_system_conf.cpp [hunk 4]
```diff
@@ -43,13 +43,85 @@ CPU::CPU() {
         return 0;
     };
 
+    auto CheckValidCpu = [&]() {
+        cpu_set_t mask;
+        CPU_ZERO(&mask);
+
+        if (sched_getaffinity(0, sizeof(cpu_set_t), &mask) == -1) {
+            return -1;
```
Comment (riverlijunjie) [inline src/inference/src/os/lin/lin_system_conf.cpp:51]: It seems we didn't use the return code.

Comment (wangleis) [reply src/inference/src/os/lin/lin_system_conf.cpp:52]: updated

File: src/inference/src/os/lin/lin_system_conf.cpp (hunk 3 above)
Comment (wangleis) [reply src/inference/src/os/lin/lin_system_conf.cpp:71]: updated

File: src/inference/src/os/lin/lin_system_conf.cpp (hunk 2 above)
Comment (wangleis) [reply src/inference/src/os/lin/lin_system_conf.cpp:100]: updated

File: src/inference/src/os/lin/lin_system_conf.cpp [hunk 5]
```diff
@@ -43,13 +43,85 @@ CPU::CPU() {
...
+                        n--;
+                    }
+                    n++;
+                }
+
+                if ((_proc_type_table.size() > 1) && (_proc_type_table[0][ALL_PROC] == _proc_type_table[1][ALL_PROC])) {
+                    _proc_type_table.pop_back();
+                }
+            }
+            _numa_nodes = _proc_type_table.size() == 1 ? 1 : _proc_type_table.size() - 1;
```
[the remark this answers is not in the document]
Comment (wangleis) [reply src/inference/src/os/lin/lin_system_conf.cpp:109]: No. `_numa_nodes = _proc_type_table.size() - 1` is used for processors on multiple sockets only.

File: src/inference/src/os/lin/lin_system_conf.cpp (hunk 1 above)
Comment (wangleis) [reply src/inference/src/os/lin/lin_system_conf.cpp]: updated.

**Review by xufang-lisa — APPROVED** [review]

File: src/inference/src/os/lin/lin_system_conf.cpp [hunk 6]
```diff
@@ -43,13 +44,88 @@ CPU::CPU() {
...
+    };
+
     if (!GetCatchInfoLinux()) {
         parse_processor_info_linux(_processors,
                                    system_info_table,
                                    _numa_nodes,
                                    _cores,
                                    _proc_type_table,
                                    _cpu_mapping_table);
+        if (CheckValidCpu()) {
```
Comment (riverlijunjie) [inline src/inference/src/os/lin/lin_system_conf.cpp:126]:
```suggestion
        if (CheckValidCpu() <0 ) {
```

Comment (wangleis) [reply src/inference/src/os/lin/lin_system_conf.cpp]: updated

File: src/inference/src/os/lin/lin_system_conf.cpp [hunk 7]
```diff
@@ -43,13 +44,88 @@ CPU::CPU() {
...
+
     if (!GetCatchInfoLinux()) {
         parse_processor_info_linux(_processors,
                                    system_info_table,
                                    _numa_nodes,
                                    _cores,
                                    _proc_type_table,
                                    _cpu_mapping_table);
+        if (CheckValidCpu() < 0) {
+            IE_THROW() << "No CPU processor available.";
```
Comment (dmitry-gorokhov) [inline src/inference/src/os/lin/lin_system_conf.cpp:127]: Can we make the message more descriptive? For now it will sounds very controversial to the user

**Review by dmitry-gorokhov — APPROVED** [review]

File: src/inference/src/os/lin/lin_system_conf.cpp (hunk 7 above)
Comment (sunxiaoxia2022) [reply src/inference/src/os/lin/lin_system_conf.cpp:127]: IE_THROW-->OPENVINO_THROW

Comment (wangleis) [reply src/inference/src/os/lin/lin_system_conf.cpp]: updated

**Pull request merged**